### PR TITLE
add AzureKeyCredentialPolicy.cs share code to eng

### PR DESCRIPTION
**Pull request by chunyu3**
# Contributing to the Azure SDK

- Generated SDK fail to compile when there is azure key authentication because `AzureKeyCredentialPolicy.cs` is shared code and internal. We need to add this to eng system, so that SDK under azure-sdk-for-net can build. 

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: eng/Directory.Build.Common.targets
Comment (christothes) [inline eng/Directory.Build.Common.targets:203]: I'm curious what motivated this change since it doesn't really simplify the inclusion of the policy. It's still a one liner in the csproj either way.

File: sdk/agrifood/Azure.Verticals.AgriFood.Farming/src/Azure.Verticals.AgriFood.Farming.csproj
Comment (annelo-msft) [inline sdk/agrifood/Azure.Verticals.AgriFood.Farming/src/Azure.Verticals.AgriFood.Farming.csproj:21]: Reiterating @christothes's question here - why is this preferred over explicitly including relevant files?  

I think the approach of defining new MSBuild properties has a lot of problems, and I would like us to move way from it.  As part of the work I was doing on https://github.com/Azure/azure-sdk-for-net/issues/28369, I was hoping to remove `<IncludeOperationsSharedSource>` eventually.  I would prefer not to be adding new properties, as this introduces technical debt that we want to eliminate eventually, so it is solving a short term problem but creating more work for us in the future.

**Review by annelo-msft — CHANGES_REQUESTED** [review]
Please do not merge without resolving https://github.com/Azure/azure-sdk-for-net/pull/39543/files#r1374780752

File: eng/Directory.Build.Common.targets
Comment (heaths) [reply eng/Directory.Build.Common.targets:203]: I also don't think we should be including it anyway. The class will end up in *every* assembly that includes shared source whether it's used or not. The C# linker doesn't elide types not used - not unless doing AOT with tree shaking.

**Review by heaths — CHANGES_REQUESTED** [review]
This unnecessarily inflates every assembly that includes shared source via the property whether they use the type or not. We should limit types included in such a way to only those used by a majority of projects.

What is the reason for doing otherwise?

**Comment by chunyu3** [comment]
> This unnecessarily inflates every assembly that includes shared source via the property whether they use the type or not. We should limit types included in such a way to only those used by a majority of projects.
> 
> What is the reason for doing otherwise?

`AzureKeyCredentialPolicy` is needed when service has azure key authentication, and because it is internal class in Azure.Core.Shared, we need to add this manually to sdk project file (.csproj) otherwise sdk compile will fail. There are several service customer met with this issue and came to us, so we decide to find a way to resolve this. 

- long-term: make `AzureKeyCredentailPolicy` public and move out Azure.Core.Shared. Azure.Core team is working on add `KeyCredentailPolicy` to system library which may can substitute this.
- Short-term workaround: add `AzureKeyCredentialPolicy` to eng as a workaround, so that every generated code can compile without manually updating the sdk project file.

File: eng/Directory.Build.Common.targets
Comment (chunyu3) [reply eng/Directory.Build.Common.targets:203]: > I'm curious what motivated this change since it doesn't really simplify the inclusion of the policy. It's still a one liner in the csproj either way.

Yes. at first workaround, we can add this policy all the time without a property switch.

**Comment by jsquire** [comment]
> * long-term: make `AzureKeyCredentailPolicy` public and move out Azure.Core.Shared. Azure.Core team is working on add `KeyCredentailPolicy` to system library which may can substitute this.
> * Short-term workaround: add `AzureKeyCredentialPolicy` to eng as a workaround, so that every generated code can compile without manually updating the sdk project file.

Why is the generator unable to add the reference in the project file automatically for those projects that have key credential auth?

**Comment by chunyu3** [comment]
> > * long-term: make `AzureKeyCredentailPolicy` public and move out Azure.Core.Shared. Azure.Core team is working on add `KeyCredentailPolicy` to system library which may can substitute this.
> > * Short-term workaround: add `AzureKeyCredentialPolicy` to eng as a workaround, so that every generated code can compile without manually updating the sdk project file.
> 
> Why is the generator unable to add the reference in the project file automatically for those projects that have key credential auth?

Yes, we can add the `AzureKeyCredentialPolicy` in project file automatically for those projects that have azure key authentication. I will implement this as our current solution. Thanks

**Pull request closed without merging**
